File: app/src/main/java/io/catter2/cat_api/FetchCatImagesUseCase.java
package io.catter2.cat_api;

import java.util.ArrayList;
import java.util.List;

public class FetchCatImagesUseCase {
    public interface Callback {
        void imagesUrls(List<String> urls);
    }

    private TheCatAPI catAPI;

    public FetchCatImagesUseCase(TheCatAPI catAPI) {
        this.catAPI = catAPI;
    }

    public void getImagesUrls(final Callback callback) {
        catAPI.getCatsWithHats(new TheCatAPI.Callback() {
            @Override
            public void response(CatImagesModel response) {
                ArrayList<String> imageUrls = new ArrayList<>();
                if (response != null) {
                    for (CatImageModel img : response.catImages) {
                        imageUrls.add(img.url);
                    }
                }
                callback.imagesUrls(imageUrls);
            }
        });
    }
}

File: app/src/main/java/io/catter2/cat_api/FetchCatImagesUseCase.kt
package io.catter2.cat_api

open class FetchCatImagesUseCase(private val catAPI: TheCatAPI) {
    open fun getImagesUrls(callback: (List<String>) -> Unit) {
        catAPI.getCatsWithHats { catImagesModel ->
            val catImagesUrls = catImagesModel
                    ?.catImages
                    ?.map { it.url }
                    ?: emptyList()

            callback(catImagesUrls)
        }
    }
}
